Stage2: Invalid execution of comptime code

### Zig Version

0.10.0-dev.3475+b3d463c9e

### Steps to Reproduce

Try compiling the following code with `zig test -fno-stage1 -fLLVM comptime-bug.zig`
```
const std = @import("std");
const S = struct { x: [4]i8 = std.mem.zeroes([4]i8) };
const M = struct { x: [4]S = std.mem.zeroes([4]S) };

comptime {
    var res = M{};
    @compileLog(res.x);
    for (.{1, 2, 3}) |i| res.x[i].x[i] = i;
    @compileLog(res.x);
}
```

### Expected Behavior

Output should be
```
comptime-bug.zig:7:5: error: found compile log statement
    @compileLog(res.x);
    ^~~~~~~~~~~~~~~~~~

Compile Log Output:
@as([4]comptime-bug.S, .{ .{.x = .{ 0, 0, 0, 0 }}, .{.x = .{ 0, 0, 0, 0 }}, .{.x = .{ 0, 0, 0, 0 }}, .{.x = .{ 0, 0, 0, 0 }} })
@as([4]comptime-bug.S, .{ .{.x = .{ 0, 0, 0, 0 }}, .{.x = .{ 0, 1, 0, 0 }}, .{.x = .{ 0, 0, 2, 0 }}, .{.x = .{ 0, 0, 0, 3 }} })
```

### Actual Behavior

Output is
```
comptime-bug.zig:7:5: error: found compile log statement
    @compileLog(res.x);
    ^~~~~~~~~~~~~~~~~~

Compile Log Output:
@as([4]comptime-bug.S, .{ .{.x = .{ 0, 0, 0, 0 }}, .{.x = .{ 0, 0, 0, 0 }}, .{.x = .{ 0, 0, 0, 0 }}, .{.x = .{ 0, 0, 0, 0 }} })
@as([4]comptime-bug.S, .{ .{.x = .{ 0, 1, 2, 3 }}, .{.x = .{ 0, 1, 2, 3 }}, .{.x = .{ 0, 1, 2, 3 }}, .{.x = .{ 0, 1, 2, 3 }} })
```

--- a/src/Sema.zig
+++ b/src/Sema.zig
@@ -23990,7 +23990,10 @@ fn beginComptimePtrMutation(
                                 const array_len_including_sentinel =
                                     try sema.usizeCast(block, src, parent.ty.arrayLenIncludingSentinel());
                                 const elems = try arena.alloc(Value, array_len_including_sentinel);
-                                mem.set(Value, elems, repeated_val);
+                                if (elems.len > 0) elems[0] = repeated_val;
+                                for (elems[1..]) |*elem| {
+                                    elem.* = try repeated_val.copy(arena);
+                                }
 
                                 val_ptr.* = try Value.Tag.aggregate.create(arena, elems);
 

--- a/test/behavior/eval.zig
+++ b/test/behavior/eval.zig
@@ -1293,3 +1293,20 @@ test "mutate through pointer-like optional at comptime" {
         try expect(payload_ptr.*.* == 16);
     }
 }
+
+test "repeated value is correctly expanded" {
+    const S = struct { x: [4]i8 = std.mem.zeroes([4]i8) };
+    const M = struct { x: [4]S = std.mem.zeroes([4]S) };
+
+    comptime {
+        var res = M{};
+        for (.{ 1, 2, 3 }) |i| res.x[i].x[i] = i;
+
+        try expectEqual(M{ .x = .{
+            .{ .x = .{ 0, 0, 0, 0 } },
+            .{ .x = .{ 0, 1, 0, 0 } },
+            .{ .x = .{ 0, 0, 2, 0 } },
+            .{ .x = .{ 0, 0, 0, 3 } },
+        } }, res);
+    }
+}

